packages/Dendrogram/src/utils/hierarchical-clustering.ts

```diff
@@ -159,3 +166,50 @@ async function hierarchicalClusteringByDistanceExec(distance: DistanceMatrix, li
 
   return newickStr;
 }
+
+function parseClusterMatrix(clusterMatrix:ClusterMatrix): NodeType {
+  /*
+  clusert matrix is in R format, I.E. the indexings are 1-based.
+  one of the reasons is that values in merge arrays are not always positive. if the value is negative
+  it means that we are referencing a leaf node. otherwise we are referencing a cluster node.
+  for example :
+  1, -2, 0.1 would mean that the merge happened between cluster 0 and leaf node 1 with a distance of 0.1
+  */
+
+  function getSubTreeLength(node: NodeType): number {
+    function subTreeLength(children?:NodeType[]): number {
+      return children && children.length ? (children[0].branch_length ?? 0) + subTreeLength(children[0].children) : 0;
+    }
+    if(isLeaf(node)) {
+      return 0;
+    } else {
+      return subTreeLength(node.children);
+    }
+  }
+
+  const clusters: NodeType[] = [];
+  const {mergeRow1, mergeRow2, heightsResult} = clusterMatrix;
+
+  for(let i = 0; i<heightsResult.length; i++) {
+    let left: NodeType, right: NodeType;
+    if(mergeRow1[i] < 0) {
+      left = {name: (mergeRow1[i] * -1 - 1).toString(), branch_length: heightsResult[i]};
+    } else {
+      left = clusters[mergeRow1[i] - 1];
+    }
+    if(mergeRow2[i] < 0) {
+      right = {name: (mergeRow2[i] * -1 - 1).toString(), branch_length: heightsResult[i]};
+    } else {
+      right = clusters[mergeRow2[i] - 1];
+    }
+
+    const leftLength = getSubTreeLength(left);
+    const rightLength = getSubTreeLength(right);
+
+    left.branch_length = heightsResult[i] - leftLength;
+    right.branch_length = heightsResult[i] - rightLength;
+
+    clusters.push({name: '', children: [left, right], branch_length: 0});
```

Comment:
better use assign by index instead of pushing, pushing is a very expensive operation
`clusters[i] = {name: '', children: [left, right], branch_length: 0};`